server/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDVersionAndSeqNoLookup.java

```diff
@@ -66,15 +66,19 @@
      */
     PerThreadIDVersionAndSeqNoLookup(LeafReader reader, String uidField) throws IOException {
         this.uidField = uidField;
-        Terms terms = reader.terms(uidField);
+        final Terms terms = reader.terms(uidField);
         if (terms == null) {
-            throw new IllegalArgumentException("reader misses the [" + uidField + "] field");
-        }
-        termsEnum = terms.iterator();
-        if (reader.getNumericDocValues(VersionFieldMapper.NAME) == null) {
-            throw new IllegalArgumentException("reader misses the [" + VersionFieldMapper.NAME + "] field");
+            // Uid is not found if a segment contains only NoOps; in this case, version should not exist either.
+            if (reader.getNumericDocValues(VersionFieldMapper.NAME) != null) {
```

Comment:
I wonder if we can make this more generic and check that the segment only has the noop fields?